api/handlers/videostream.go

```diff
@@ -35,54 +35,350 @@ LICENSE
 func GetVideoStreams(ctx *fiber.Ctx) error {
+	// Parse URL.
 	qry := new(GetVideoStreamsQuery)
 	qry.SetLimit()
 
 	if err := ctx.QueryParser(qry); err != nil {
-		return api.InvalidRequestURL(ctx)
+		return api.InvalidRequestURL()
 	}
 
-	// Placeholder code: returns an empty result.
-	// TODO: implement fetching from datastore.
-	result := api.Result[VideoStreamResult]{
-		Results: []VideoStreamResult{},
+	format := new(api.Format)
+	if err := ctx.QueryParser(format); err != nil {
+		return api.InvalidRequestURL()
+	}
+
+	// Fetch data from the datastore.
+	store := ds_client.Get()
+	query := store.NewQuery("VideoStream", false)
+
+	// Filter by timespan.
+	if qry.TimeSpan != nil {
+		if qry.TimeSpan.Start.After(qry.TimeSpan.End) { // Validate timespan
+			return api.InvalidRequestURL()
+		}
+
+		query.FilterField("StartTime", "<=", qry.TimeSpan.End)
+		query.FilterField("EndTime", ">=", qry.TimeSpan.Start)
+	}
+
+	// Filter by capture source.
+	if qry.CaptureSource != nil {
+		query.FilterField("CaptureSource", "=", *qry.CaptureSource)
+	}
+
+	query.Limit(qry.Limit)
+	query.Offset(qry.Offset)
+
+	var videoStreams []model.VideoStream
+	keys, err := store.GetAll(context.Background(), query, &videoStreams)
+	if err != nil {
+		return api.DatastoreReadFailure()
+	}
+
+	// Format results.
+	results := make([]VideoStreamResult, len(videoStreams))
+	for i := range videoStreams {
+		results[i] = FromVideoStream(&videoStreams[i], int(keys[i].ID), format)
+	}
+
+	return ctx.JSON(api.Result[VideoStreamResult]{
+		Results: results,
 		Offset:  qry.Offset,
 		Limit:   qry.Limit,
-		Total:   0,
+		Total:   len(results),
+	})
+}
+
+// CreateVideoStream creates a new video stream.
+// BUG: start and end time are required but this is not being enforced.
+func CreateVideoStream(ctx *fiber.Ctx) error {
+
+	// Parse body.
+	var body CreateVideoStreamBody
+
+	err := ctx.BodyParser(&body)
+	if err != nil {
+		return api.InvalidRequestJSON()
 	}
-	return ctx.JSON(result)
+
+	// Create video stream entity and add to the datastore.
+	cs := model.VideoStream{
+		StartTime:     body.StartTime,
+		EndTime:       &body.EndTime,
+		StreamUrl:     body.StreamUrl,
+		CaptureSource: body.CaptureSource,
+	}
+
+	// Verify CaptureSource exists.
+	store := ds_client.Get()
+	key := store.IDKey("CaptureSource", body.CaptureSource)
+	var captureSource model.CaptureSource
+	if store.Get(context.Background(), key, &captureSource) != nil {
+		return api.DatastoreReadFailure()
+	}
+
+	// Get a unique ID for the new video stream.
+	key = store.IncompleteKey("VideoStream")
+
+	// Add to datastore.
+	key, err = store.Put(context.Background(), key, &cs)
+	if err != nil {
+		print(err.Error())
+		return api.DatastoreWriteFailure()
+	}
+
+	// Return ID of created video stream.
+	id := int(key.ID)
+	return ctx.JSON(VideoStreamResult{
+		ID: &id,
+	})
+}
+
+// StartVideoStream creates a new video stream at the current time.
+func StartVideoStream(ctx *fiber.Ctx) error {
```

Comment:
This function looks extremely similar to the one above. It's probably possible to factor out the common stuff into a function and then have these wrap that ?